### PR TITLE
Fixed memoization of RankedModel::Ranker::Mapper#finder

### LINKED ISSUE #103: Moving last item up reverses the order of whole scope when 3 last items have adjacent indexes
If order index for the last items are adjacent (e.g. last => 15, last-1 => 14, last-2 => 13), attempting to move the last item up causes the whole scope to be reordred in reverse. For instance:

```
id  row_order   scope_value
1   9760    1
2   9762    1
3   9772    1
4   9773    1
5   9774    1
```

``` ruby
Model.find(5).update_attribute(:row_oder, :up)
```

gives

```
id  row_order   scope_value
5   -5592404    1
4   -2796202    1
3   0   1
2   2796203 1
1   5592405 1
```

### PR DESCRIPTION
**Pull request by brendon**
It now caches a value for each passed sort order; without this multiple calls to finder can cause weird flipping issues where the whole list gets reversed.

NOTE: @mixonic, this is basically the same as https://github.com/mixonic/ranked-model/pull/110 but with an associated test that fails without the fix. The test came from here: https://github.com/mixonic/ranked-model/commit/37e19f899119dd5825299c8611fff75be1c631a9

Credit to @jmchambers and @mpokrywka for the actual work.

Closes #103 

Also makes https://github.com/mixonic/ranked-model/pull/110 redundant.

### CODE REVIEW

**Comment by brendon** [comment]
The automated test suite is a bit out of date (and thus failing to install in some instances). I've successfully run the tests locally with the `sqlite` driver on Active Record '4.2'.

**Comment by brendon** [comment]
@mixonic, if you're stretched for time, would you consider adding me as a maintainer? I currently maintain `acts_as_list` with @swanandp and he can vouch for me.

**Comment by swanandp** [comment]
FWIW, @brendon has my full endorsement, I am thankful for all the great work he's done on `acts_as_list`.

**Comment by brendon** [comment]
@mixonic, what are the chances of getting this merged? If I put in the work to get the test suite installing correctly, will you have the time to merge?

**Comment by mixonic** [comment]
@brendon yes of course, and I'm happy to add you as a maintainer as well.

I'd very much like to minimize merging failing builds, or at least only merge them if we have a documented plan to resolve the issues quickly. In the supported Travis matrix only builds of 1.9.3 and rbx are failing.

**Comment by mixonic** [comment]
For Ruby 1.9.3 it looks like we need to pin the version of `pg`.

screen shot 2018-07-09 at 5 37 37 pm

**Comment by brendon** [comment]
Thanks @mixonic, I've used Appraisal in the past to better facilitate the travis test matrix. I'm attempting to convert `ranked-model` to that at the moment. Will post back once it's green :)

**Comment by brendon** [comment]
Lol, that's the first conversion I've done that's been green first go! :D https://travis-ci.org/mixonic/ranked-model/builds/402557044

I excluded rubinius and jruby. I'll now add them back in. Depending on the results, perhaps we should exclude rubinius?

**Comment by brendon** [comment]
Looks like we have a flappy test:

https://travis-ci.org/mixonic/ranked-model/jobs/402589391#L1133
https://travis-ci.org/mixonic/ranked-model/jobs/402584912#L940

For now I guess let's ignore that and get this PR merged. I'm close to getting it green now for all rubies. Just waiting on a green for rbx and jruby, then will merge them all together and do one last run. Hopefully our flappy test doesn't rear its head.

**Comment by brendon** [comment]
Looking a bit closer at the test, could you explain it for me? I've open a seperate issue: https://github.com/mixonic/ranked-model/issues/126

**Comment by brendon** [comment]
Hi @mixonic :) Great to see the matrix is green! I think merge this, then we can look at why that test is occasionally failing. We can address it in another PR.

And yes please, if you can make me a maintainer that would be great. I'll then try to get some action on the other PR's and issues waiting in the wings.

**Comment by brendon** [comment]
@mixonic, can you force a re-run of the suite? One of the VM's stalled:

https://travis-ci.org/mixonic/ranked-model/jobs/403051387#L447

File: .travis.yml [hunk 1]
```diff
@@ -1,68 +1,38 @@
...
-  - "2.2.4"
-  - "2.3.0"
-  - "jruby-19mode"
-  - "rbx"
+  - 1.9.3
+  - 2.1.10
+  - 2.2.10
+  - 2.3.7
+  - 2.4.4
+  - 2.5.1
```
Comment (mixonic) [inline .travis.yml]: Fantastic! Thank you for updating this.

File: spec/duck-model/lots_of_ducks_spec.rb [hunk 2]
```diff
@@ -93,8 +93,8 @@
         Duck.first(50).each_with_index do |d, index|
           d.update_attributes :age => index % 10, :pond => "Pond #{index / 10}"
         end
-        @duck_11 = Duck.offset(10).first
-        @duck_12 = Duck.offset(11).first
+        @duck_11 = Duck.where(:pond => 'Pond 1').rank(:age).first
+        @duck_12 = Duck.where(:pond => 'Pond 1').rank(:age).second
```
Comment (mixonic) [inline spec/duck-model/lots_of_ducks_spec.rb]: Seems ok. I'm very curious [why `offset` isn't sufficient](https://github.com/mixonic/ranked-model/issues/126#issuecomment-404571071).

**Comment by brendon** [comment]
Thanks @mixonic! :) Great to have this merged :D

My next job will be looking at Rails 5.x support. I don't think there's too much to do (some stuff around `Integer`) but I'll create a PR with Rails 5 in the test suite and go from there.

Thanks again!

**Comment by mixonic** [comment]
landed in v0.4.1

**Pull request merged**